gateway/handler.go

```diff
@@ -335,6 +351,61 @@ func (h *CORSHandler) Shutdown(ctx context.Context) error {
 	return shutdown(ctx, h.next)
 }
 
+// LoggingHandler logs incoming HTTP requests with details
+type LoggingHandler struct {
+	next http.Handler
+}
+
+// NewLoggingHandler creates a new LoggingHandler that logs request details
+func NewLoggingHandler(next http.Handler) *LoggingHandler {
+	return &LoggingHandler{next: next}
+}
+
+func (h *LoggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Prepare log fields
+	logFields := []interface{}{
+		"remote_ip", getRemoteIP(r),
+		"method", r.Method,
+		"url", r.URL.String(),
+	}
+
+	// For POST requests, try to read and log up to maxLogBodyBytes of the body
+	const maxLogBodyBytes = 1024
+	if r.Method == http.MethodPost {
+		limited := &io.LimitedReader{R: r.Body, N: maxLogBodyBytes + 1}
+		buf, err := io.ReadAll(limited)
+		if err == nil {
+			var bodyStr string
+			if int64(len(buf)) > maxLogBodyBytes {
+				bodyStr = string(buf[:maxLogBodyBytes]) + "...[truncated]"
+			} else {
+				bodyStr = string(buf)
+			}
+			logFields = append(logFields, "body", bodyStr)
+			// Reconstruct the body for downstream handlers: combine what we read and the rest
+			rest := io.MultiReader(bytes.NewReader(buf), r.Body)
+			r.Body = io.NopCloser(rest)
+		}
+	}
+
+	log.Infow("request", logFields...)
+
+	h.next.ServeHTTP(w, r)
```

Comment:
What comes `next` after this handler? I don't think the next handler will have access to the full body since the first 1024 bytes are read here.